src/KubernetesClient/PeekableStreamReader.cs

```diff
@@ -0,0 +1,118 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace k8s
+{
+    public class PeekableStreamReader : IAsyncLineReader
+    {
+        private readonly int _maxLineLength;
+        private readonly Queue<string> _peek = new Queue<string>();
+        private Stream _stream;
+        private byte[] _buf = new byte[4096];
+        private int _bufPos = 0;
+        private int _bufLength = 0;
+        private bool _eof = false;
+
+        public PeekableStreamReader(Stream stream, int maxLineLength = Int32.MaxValue)
+        {
+            _maxLineLength = maxLineLength;
+            _stream = stream;
+        }
+
+        private void ShuffleConsumedBytes(byte[] dst)
+        {
+            if (_bufPos < _bufLength)
+            {
+                Buffer.BlockCopy(_buf, _bufPos, dst, 0, _bufLength - _bufPos);
+            }
+            _bufLength -= _bufPos;
+            _bufPos = 0;
+        }
+
+        private void CheckDisposed()
+        {
+            if (_stream == null)
+            {
+                throw new ObjectDisposedException("PeekableStreamReader");
+            }
+        }
+
+        private async Task<string> ReadLineNoPeekAsync(CancellationToken cancellationToken)
+        {
+
+            for (; ; )
+            {
+                // read buffered data
+                if (_bufPos < _bufLength)
+                {
+                    int nlPos = _eof ? _bufLength - 1 : Array.IndexOf(_buf, (byte)'\n', _bufPos, _bufLength - _bufPos);
+                    if (nlPos >= 0)
+                    {
+                        string result = Encoding.UTF8.GetString(_buf, _bufPos, nlPos - _bufPos + 1);
+                        _bufPos = nlPos + 1;
+                        return result;
+                    }
+                }
+                if (_eof)
+                {
+                    return null;
+                }
+                // consume any previously read data
+                if (_bufPos > 0)
+                {
+                    ShuffleConsumedBytes(_buf);
+                }
+                // make the buffer bigger if needed
+                if (_buf.Length - _bufLength < 512)
+                {
+                    if (_buf.Length > _maxLineLength / 2)
+                    {
+                        throw new InvalidOperationException(
+                            "trying to read a line > 2Gb from the server");
```

Comment:
Additional info could be helpful here. Perhaps adding a small sentence that clarifies that such sizes are not supported.

Reply:
I have updated this to read:

```csharp
throw new InvalidOperationException(
    "Reading from the stream failed because a line from the server was unexpectedly " +
    $"greater than the configured maximum line length of {_maxLineLength}");
```

Reply:
I have changed this to read:

```csharp
throw new InvalidOperationException(
    "Reading from the stream failed because a line from the server was unexpectedly " +
    $"greater than the configured maximum line length of {_maxLineLength}");
```